### ISSUE TITLE
Run zookeeper mainloop in a separate thread

### ISSUE THREAD

**Issue by bitchkat** [shayhatsor/zookeeper#29]
I'm seeing some zookeeper disconnects that is almost surely because business logic is running for a significant amount of time in synchronous code.  Ideally, I would like to avoid application developers from having to go down the async rabbit hole if possible.

This is also an issue when I'm stopped in a debugger as it will likely disconnect.  

It seems that it should be possible so that zookeeper's main loop is run in a dedicated thread so that it can send it heartbeats to zookeeper server even if the application is blocking in synchronous code.

Are there any recommendations or best practices to achieve this?

**Comment by shayhatsor** [comment]
this is probably related to #28. I think that the problem isn't with having app developers write async code, it's with enqueuing long running code chunks on the thread pool. Which in practice causes starvation to other thread pool queued code. The solution I suggest is to run the long running code on a dedicated thread instead of the ZK client. This is an async client, it's built on cooperative multi tasking. To make it run on dedicated threads would mean rewriting the code to not use tasks at all. Or implement a custom TaskScheduler and remove all `ConfigureAwait(false)`s. Or use explicit threads with tasks ? i have no idea if that even works.

**Comment by bitchkat** [comment]
I actually did some work on this.  My theory was to run the Zookeeper mainloop in a dedicated thread.  I made the following changes:

- Created a custom scheduler to run startTask and eventTask at highest priority

- Changed ConfigureAwait(false) to ConfgureAwait(true) in ClientCnxn.cs and ClientCnxnSocketNIO.cs

- In doTransport(), subvert the call the Task.WhenAny if somethingPending.Task is completed.  Otherwise, the WhenAny was yielding the CPU and the task could experience significant delay in getting CPU time slice again.

My test program consists of a configurable number of tasks that just spin in a loop for much longer than the ZK timeout (30 seconds).  With those changes above, the zk pings and responses are occurring in a timely manner and not getting starved for CPU.  I still need to do a lot of testing to ensure there are no unexpected consequences.

My reasoning for this approach is that I can control how the infrastructure works but I can't as easily control what application developers do.  As much as possible, I'd like to keep business logic unaware that its running on a platform with liveness responsibilities.
